File: c/Problem3.c
/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 *
 * What is the largest prime factor of the number 600851475143 ?
 */
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

typedef void lfactorize_callback_t(long factor, void *context);

void lfactorize(long x, lfactorize_callback_t *callback, void *context) {
    long divisor = 2;
    while (x >= divisor) {
        ldiv_t dm = ldiv(x, divisor);
        if (dm.rem) {
            divisor++;
        } else {
            callback(divisor, context);
            x = dm.quot;
        }
    }
}

static void lfactorize_callback(long factor, void *context) {
    if (context) {
        *(long *)context = factor;
    }
}

int main(void) {
    long max_factor;
    lfactorize(600851475143, &lfactorize_callback, &max_factor);
    printf("%ld\n", max_factor);
    return 0;
}

File: rust/Problem3.rs
/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 *
 * What is the largest prime factor of the number 600851475143 ?
 */
use std::collections::HashMap;

fn factorize(x: i64) -> HashMap<i64, i64> {
    let mut map: HashMap<i64, i64> = HashMap::new();
    let mut x = x;
    let mut divisor = 2;
    while x >= divisor {
        let rem = x % divisor;
        if rem > 0 {
            divisor += 1;
        } else {
            let counter = map.entry(divisor).or_insert_with(|| 0);
            *counter += 1;
            x /= divisor;
        }
    }
    map
}

fn main() {
    let map = factorize(600851475143);
    println!("{}", map.keys().max().unwrap());
}
